cibuildwheel/options.py

```diff
@@ -137,32 +131,31 @@ def __call__(
         table: Optional[TableFmt] = None,
     ) -> str:
         """
-        Get and return envvar for name or the override or the default. If
-        env_plat is False, then don't accept platform versions of the
-        environment variable. If this is an array it will be merged with "sep"
-        before returning. If it is a table, it will be formatted with
-        "table['item']" using {k} and {v} and merged with "table['sep']".
+        Get and return the value for the named option from environment,
+        configuration file, or the default. If env_plat is False, then don't
+        accept platform versions of the environment variable. If this is an
+        array it will be merged with "sep" before returning. If it is a table,
+        it will be formatted with "table['item']" using {k} and {v} and merged
+        with "table['sep']".
         """
 
-        if name not in self.config:
+        if name not in self.default_options and name not in self.default_platform_options:
             raise ConfigOptionError(f"{name} must be in cibuildwheel/resources/defaults.toml file")
 
         # Environment variable form
-        envvar = f"CIBW_{name.upper().replace('-', '_').replace('.', '_')}"
+        envvar = f"CIBW_{name.upper().replace('-', '_')}"
         plat_envvar = f"{envvar}_{self.platform.upper()}"
 
-        # Let environment variable override setting in config
-        if env_plat:
-            result = _dig_first(
-                (os.environ, plat_envvar),
-                (os.environ, envvar),
-                (self.config, name),
-            )
-        else:
-            result = _dig_first(
-                (os.environ, envvar),
-                (self.config, name),
-            )
+        # get the option from the environment, then the config file, then finally the default.
+        # platform-specific options are preferred, if they're allowed.
+        result = _dig_first(
+            (os.environ if env_plat else {}, plat_envvar),  # type: ignore
```

Comment:
Curious, `os.environ if env_plat else {}` types as `object`, rather than a mapping. But individually at least `os.environ` works.

Reply:
yeah, I didn't understand that. I might have expected Union[_Environ, dict]. Actually that's what pyright gives for the expression. I don't know enough about types to know if that's a mypy bug or not!